cpp/arcticdb/storage/rocksdb/rocksdb_storage.cpp

```diff
@@ -0,0 +1,239 @@
+        }
+    });
+}
+
+bool RocksDBStorage::do_key_exists(const VariantKey& key) {
+    ARCTICDB_SAMPLE(RocksDBStorageKeyExists, 0)
+    std::string value; // unused
+    auto key_type_name = fmt::format("{}", variant_key_type(key));
+    auto k_str = to_serialized_key(key);
+    auto s = db_->Get(::rocksdb::ReadOptions(), handles_by_key_type_.at(key_type_name), ::rocksdb::Slice(k_str), &value);
+    return !s.IsNotFound();
+}
+
+void RocksDBStorage::do_remove(Composite<VariantKey>&& ks, RemoveOpts opts)
+{
+    ARCTICDB_SAMPLE(RocksDBStorageRemove, 0)
+
+    auto failed_deletes = do_remove_internal(std::move(ks), opts);
+    if (!failed_deletes.empty()) {
+        throw KeyNotFoundException(Composite<VariantKey>(std::move(failed_deletes)));
+    }
+}
+
+bool RocksDBStorage::do_fast_delete() {
+    foreach_key_type([&](KeyType key_type) {
+        if (key_type == KeyType::TOMBSTONE) {
+            // TOMBSTONE and LOCK both format to code 'x' - do not try to drop both
+            return;
+        }
+        auto key_type_name = fmt::format("{}", key_type);
+        auto handle = handles_by_key_type_.at(key_type_name);
+        ARCTICDB_DEBUG(log::storage(), "dropping {}", key_type_name);
+        db_->DropColumnFamily(handle);
+    });
+    return true;
+}
+
+void RocksDBStorage::do_iterate_type(KeyType key_type, const IterateTypeVisitor& visitor, const std::string& prefix) {
+    ARCTICDB_SAMPLE(RocksDBStorageItType, 0)
+    auto prefix_matcher = stream_id_prefix_matcher(prefix);
+
+    auto key_type_name = fmt::format("{}", key_type);
+    auto handle = handles_by_key_type_.at(key_type_name);
+    ::rocksdb::Iterator* it = db_->NewIterator(::rocksdb::ReadOptions(), handle);
+    for (it->SeekToFirst(); it->Valid(); it->Next()) {
+        auto key_slice = it->key();
+        auto k = variant_key_from_bytes(reinterpret_cast<const uint8_t *>(key_slice.data()), key_slice.size(), key_type);
+
+        ARCTICDB_DEBUG(log::storage(), "Iterating key {}: {}", variant_key_type(k), variant_key_view(k));
+        if (prefix_matcher(variant_key_id(k))) {
+            visitor(std::move(k));
+        }
+    }
+    auto s = it->status();
+    util::check(s.ok(), DEFAULT_ROCKSDB_NOT_OK_ERROR + s.ToString());
+    delete it;
+}
+
+std::vector<VariantKey> RocksDBStorage::do_remove_internal(Composite<VariantKey>&& ks, RemoveOpts opts)
+{
+    auto grouper = [](auto &&k) { return variant_key_type(k); };
+    std::vector<VariantKey> failed_deletes;
+
+    (fg::from(ks.as_range()) | fg::move | fg::groupBy(grouper)).foreach([&](auto &&group) {
+        auto key_type_name = fmt::format("{}", group.key());
+        // If no key of this type has been written before, this can fail
+        auto handle = handles_by_key_type_.at(key_type_name);
+        auto options = ::rocksdb::WriteOptions(); // Should this be const class attr? Used in write_internal too.
+        options.sync = true;
+        for (const auto &k : group.values()) {
+            if (do_key_exists(k)) {
```

Comment:
Would it not be more efficient to try to delete it no matter what, and just do the error handling logic you have already if the status after the delete is not OK?

Reply:
Unfortunately RocksDB exists silently with OK if the key we are trying to delete does not exist. I think LMDB behaves differently, for instance. We need to be able to report back of any failed deletes, so this is necessary.